feat: Add new azure_imds node_attestor plugin

0xRBell opened the pull request:
<!--
1. If this is your first PR, please read our contributor guidelines
https://github.com/spiffe/spiffe/blob/master/CONTRIBUTING.md
https://github.com/spiffe/spire/blob/main/CONTRIBUTING.md

2. Please remember to include a DCO on every commit (`git commit -s`)
https://github.com/apps/dco
-->

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**
Adds a new mechanism for attesting VMs and VM Scale sets in azure.

**Description of change**
This plugin ,`azure_imds`,  uses the new azure IMDS attested document instead of the older MSI Token in `azure_msi`. 
Key take aways:

1. Adds attestation for VM Scale sets(Flexible and Uniform)
2. Adds additional selectors such as tags, location, vm scale set name. (maintains all the selectors of `azure_msi`
3. Uses IMDS attested document to obtain vmID and subscription id.
4. No longer requires a VM to have a managed identity assigned as it no longer relies on the MSI Token.
5. Adds a new authentication mechanism: JWT Token federation to an AppReg 
6. Allows for attesting nodes from multiple subscriptions in a single tenant, as opposed to a single subscription per tenant as is the case in `azure_msi`

JWT Federation Auth:
This new mechanism works by creating a credential assertion that reads a JWT from disk. This allows for easy integration with things such as K8S project service account tokens which mount a JWT to disk and keep it updated.

new imported library [smallstep/pkcs7](https://github.com/smallstep/pkcs7) was needed for pkcs7 signature verification as it contains updates to work with Azure signature algorithm combinations. This is a for of [mozilla](https://github.com/mozilla-services/pkcs7) which itself is a fork of [fullsailor](https://github.com/fullsailor/pkcs7) which is currently used in Spire but has not been updated in 6 years.

**Which issue this PR fixes**
fixes https://github.com/spiffe/spire/issues/6268
fixes https://github.com/spiffe/spire/issues/2527

0xRBell committed:
feat: Add new azure_imds plugin

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
clean up readme, add more selectors

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
cleanup

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
refactor vmss instance interfaces

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
cleanup

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
update: this commit gets the plugin into a working state. successfully attested azure nodes in a vm scale set

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
remove playground testing file

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
use smallstep for pkcs7, its the only one that has been updated to support azure documents

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
add certificate validation

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
cleanup comments

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
cleanup commens and print statements

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
require nonce from server to be in the attested document

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

amartinezfayo reviewed (commented):
Thank you so much @0xRBell for this contribution, this is looking great!
I've added a lot of minor comments, but also some comments related wit concerns  around validation logic.

0xRBell committed:
Update pkg/agent/plugin/nodeattestor/azureimds/imds_test.go

Co-authored-by: Agustín Martínez Fayó <amartinezfayo@gmail.com>
Signed-off-by: Ryan <bellrd07@gmail.com>
Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
Update pkg/server/plugin/nodeattestor/azureimds/doc_validation.go

Co-authored-by: Agustín Martínez Fayó <amartinezfayo@gmail.com>
Signed-off-by: Ryan <bellrd07@gmail.com>
Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell commented on pkg/server/plugin/nodeattestor/azureimds/utils.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+func generateRandomAlphanumeric(length int) (string, error) {
+	const alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
```
I can make it 32 bytes long which I think is the max and it only accepts alpha numeric characters.

0xRBell commented on pkg/server/plugin/nodeattestor/azureimds/doc_validation.go:
```diff
@@ -0,0 +1,218 @@
+		return fmt.Errorf("certificate has no common name in subject")
+	}
+
+	if !strings.Contains(subject, expectedSubject) {
```
yes I think exact match should be ok. I originally did a contains to try and allow either of the Digicert Root CAs. The older that ends with CA and the newer that ends with G2. I can switch this to exact match though

0xRBell commented on pkg/server/plugin/nodeattestor/azureimds/doc_validation.go:
```diff
@@ -0,0 +1,218 @@
+	}
+
+	// Fetch the intermediate certificate
+	req, err := http.NewRequestWithContext(ctx, "GET", caIssuersURL, nil)
```
will update to set a timeout

0xRBell committed:
Update pkg/server/plugin/nodeattestor/azureimds/imds.go

Co-authored-by: Agustín Martínez Fayó <amartinezfayo@gmail.com>
Signed-off-by: Ryan <bellrd07@gmail.com>
Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell commented on doc/plugin_server_nodeattestor_azure_imds.md:
```diff
@@ -0,0 +1,217 @@
+| Field        | Required | Description                   |
+| ------------ | -------- | ----------------------------- |
+| `app_id`     | Required | The application (client) ID   |
+| `token_path` | Required | Path on disk to the JWT token |
```
that makes sense, I will update that name

0xRBell commented on pkg/server/plugin/nodeattestor/azureimds/doc_validation.go:
```diff
@@ -0,0 +1,218 @@
+		}
+	}
+
+	opts := x509.VerifyOptions{
```
I will look into the .wellknown, if not I can add the constants since there is already prior art for that.

0xRBell committed:
bump nonce to 32 bytes
hard code root ca
update issuer and subject to exact match
add timeouts to http request calls

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>
Signed-off-by: Ryan <bellrd07@gmail.com>

0xRBell commented on pkg/server/plugin/nodeattestor/azureimds/doc_validation.go:
```diff
@@ -0,0 +1,218 @@
+		return nil, fmt.Errorf("failed to create request for intermediate certificate: %w", err)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
```
the issuer URL from the cert is 
`http://www.microsoft.com/...` 
I can check that its from microsoft but it looks like its not an https url

0xRBell committed:
add more issuer url validation for intermediate

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
remove debug log from catalog

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

amartinezfayo reviewed (commented):
Thank you @0xRBell for the updates.
I've added some more comments.

0xRBell commented on pkg/server/plugin/nodeattestor/azureimds/client.go:
```diff
@@ -0,0 +1,306 @@
+	query := fmt.Sprintf(`
+	resources 
+	| where type =~ 'microsoft.compute/virtualmachines'
+	| where properties.vmId == '%s'
```
The vmId comes out of the attested document so it _should_ always be valid. But I can definitely add some logic just to make sure.

0xRBell commented on pkg/server/plugin/nodeattestor/azureimds/doc_validation.go:
```diff
@@ -0,0 +1,242 @@
+	"net/url"
+	"time"
+
+	"github.com/smallstep/pkcs7"
```
I added some comments in the description of this PR about why this is needed. The other libraries do not support the signature/algo combo that azure uses so we either need to upstream the changes that smallstep added to support it or just use smallstep

0xRBell committed:
add tests for utils

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
add more test

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
add tests for agent and common

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
udpate documentation

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
clean up go.mod

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell commented on pkg/server/plugin/nodeattestor/azureimds/client.go:
```diff
@@ -0,0 +1,306 @@
+	query := fmt.Sprintf(`
+	resources 
+	| where type =~ 'microsoft.compute/virtualmachinescalesets'
+	| where name == '%s'
```
added validation for scales set name

0xRBell committed:
update doc validation to use regex for issuers instead of string match

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

amartinezfayo commented on pkg/server/plugin/nodeattestor/azureimds/imds.go:
```diff
@@ -0,0 +1,416 @@
+	case docData.VMID == "":
+		return status.Errorf(codes.InvalidArgument, "missing VM ID in attested document")
+	case validateUUID(docData.VMID) != nil:
+		return status.Errorf(codes.InvalidArgument, "invalid VM ID: %v", err)
```
The err variable here is not related with this error. I think that this should be something like this:

```suggestion
	case err := validateUUID(docData.VMID); err != nil:
		return status.Errorf(codes.InvalidArgument, "invalid VM ID: %v", err)

```

amartinezfayo reviewed (commented):
Thanks for addressing the comments, @0xRBell!
Please note that some tests are failing.

0xRBell committed:
update test

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
change config name

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
implement restrict_to_subscriptions

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell commented on pkg/agent/plugin/nodeattestor/azureimds/imds.go:
```diff
@@ -0,0 +1,157 @@
+	return &configv1.ConfigureResponse{}, nil
+}
+
+func (p *IMDSAttestorPlugin) Validate(_ context.Context, req *configv1.ValidateRequest) (*configv1.ValidateResponse, error) {
```
I had commited before i saved the tests last time so several were completely broken which caused the failures. I have updated them and also added one to cover Validate you mentioned above

0xRBell committed:
update go files

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>
Signed-off-by: Ryan <bellrd07@gmail.com>

0xRBell committed:
reset go mod and sum

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>
Signed-off-by: Ryan <bellrd07@gmail.com>

amartinezfayo commented:
@0xRBell I see that you made a commit, but it seems like the DCO and lint issues are not fixed. Could you look at that? We are planning to include this in the upcoming 1.14.0 release.

0xRBell committed:
fix linting issues

Signed-off-by: Ryan Bell <ryan.bell@tanium.com>

0xRBell committed:
go mod tidy

Signed-off-by: Ryan <bellrd07@gmail.com>

amartinezfayo reviewed (approved):
Thank you @0xRBell for this contribution! 🎉

amartinezfayo committed:
Merge branch 'main' into azure-imds

sorindumitru committed:
Merge branch 'main' into azure-imds

MikeZappa87 commented on pkg/server/plugin/nodeattestor/azureimds/roots.go:
```diff
@@ -0,0 +1,60 @@
+-----END CERTIFICATE-----`
+)
+
+var roots = []string{DigiCertGlobalRootG2Pem, DigiCertGlobalRootCAPem}
```
@0xRBell kinda late to the party here, I was curious why these are hardcoded? Shouldn't azure VM's have these root ca in the /etc/ssl/certs directory or in the apt ca-certificates pkg?

0xRBell replied:
Depends where you're running the server from. If its running in a container built off SCRATCH then it won't have any of those. This also followed the same pattern established in the aws_iid node attestor.

There may also be an argument that if you don't hard code the roots someone could technically be trusted if the were able to sneak their certs into the list of trusted CAs
